watchtower/wtdb/tower_db.go

```diff
@@ -0,0 +1,577 @@
+// method will fail to prevent accidental reversions.
+func (t *TowerDB) syncVersions(versions []version) error {
+	curVersion, err := t.Version()
+	if err != nil {
+		return err
+	}
+
+	latestVersion := getLatestDBVersion(versions)
+	switch {
+
+	// Current version is higher than any known version, fail to prevent
+	// reversion.
+	case curVersion > latestVersion:
+		return channeldb.ErrDBReversion
+
+	// Current version matches highest known version, nothing to do.
+	case curVersion == latestVersion:
+		return nil
+	}
+
+	// Otherwise, apply any migrations in order to bring the database
+	// version up to the highest known version.
+	updates := getMigrations(versions, curVersion)
+	return t.db.Update(func(tx *bbolt.Tx) error {
+		for _, update := range updates {
+			if update.migration == nil {
+				continue
+			}
+
+			log.Infof("Applying migration #%d", update.number)
+
+			err := update.migration(tx)
+			if err != nil {
+				log.Errorf("Unable to apply migration #%d: %v",
+					err)
+				return err
+			}
+		}
+
+		return putDBVersion(tx, latestVersion)
+	})
+}
+
+// Version returns the database's current version number.
+func (t *TowerDB) Version() (uint32, error) {
+	var version uint32
+	err := t.db.View(func(tx *bbolt.Tx) error {
+		var err error
+		version, err = getDBVersion(tx)
+		return err
+	})
+	if err != nil {
+		return 0, err
+	}
+
+	return version, nil
+}
+
+// Close closes the underlying database.
+func (t *TowerDB) Close() error {
+	return t.db.Close()
+}
+
+// GetSessionInfo retrieves the session for the passed session id. An error is
+// returned if the session could not be found.
+func (t *TowerDB) GetSessionInfo(id *SessionID) (*SessionInfo, error) {
+	var session *SessionInfo
+	err := t.db.View(func(tx *bbolt.Tx) error {
+		sessions := tx.Bucket(sessionsBkt)
+		if sessions == nil {
+			return ErrUninitializedDB
+		}
+
+		var err error
+		session, err = getSession(sessions, id[:])
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return session, nil
+}
+
+// InsertSessionInfo records a negotiated session in the tower database. An
+// error is returned if the session already exists.
+func (t *TowerDB) InsertSessionInfo(session *SessionInfo) error {
+	return t.db.Update(func(tx *bbolt.Tx) error {
+		sessions := tx.Bucket(sessionsBkt)
+		if sessions == nil {
+			return ErrUninitializedDB
+		}
+
+		sessionBytes := sessions.Get(session.ID[:])
+		if sessionBytes != nil {
+			return ErrSessionAlreadyExists
+		}
+
+		return putSession(sessions, session)
+	})
+}
+
+// InsertStateUpdate stores an update sent by the client after validating that
+// the update is well-formed in the context of other updates sent for the same
+// session. This include verifying that the sequence number is incremented
+// properly and the last applied values echoed by the client are sane.
+func (t *TowerDB) InsertStateUpdate(update *SessionStateUpdate) (uint16, error) {
+	var lastApplied uint16
+	err := t.db.Update(func(tx *bbolt.Tx) error {
```

Comment:
`Batch`?

Reply:
Applies to write transactions in the entire file as well.

Reply:
I think we should hold off on adding `Batch` just yet. We have to consider that the inputs are adversarial, and specially crafted inputs can cause the methods below to produce errors and induce even more latency since everything in the batch must be replayed (and this method will execute again in isolation). To do batching properly, we would want `Update` return nil in all error cases that could be adversarially triggered, and then capture the specific error in the surrounding scope. This allows `Batch` to see the operations as successful, even if they trigger no db mutations, but still return the appropriate error to the caller

Reply:
Can't the same be said for any node announcement gossip message? Rationale behind optimistically batching is to ensure a tower can keep up with the level of traffic at scale. We'll either need to do it here, do it at a higher level (coalesce several updates within the server), or switch to something like `leveldb` which is write optimized.